File: frontend/src/api/automation.ts
/**
 * Automation API — 自动化规则管理
 */
import api from './index'

export interface AutomationRule {
  id: number
  name: string
  description: string
  project_id: number
  trigger_type: string
  conditions: string
  actions: string
  is_enabled: boolean
  sequence: number
  execution_count: number
  created_at: string
  updated_at: string
}

export interface AutomationExecution {
  id: number
  rule_id: number
  issue_id: number
  trigger_type: string
  context_json: string
  actions_taken: string
  status: string
  error: string
  duration: number
  executed_at: string
}

export interface AutomationExecutionResponse {
  data: AutomationExecution[]
  total: number
  limit: number
  offset: number
}

export interface AutomationCreate {
  name: string
  description?: string
  trigger_type: string
  conditions?: string
  actions: string
  is_enabled?: boolean
  sequence?: number
}

export const automationApi = {
  list: async (projectId: number): Promise<AutomationRule[]> => {
    const res = await api.get(`/projects/${projectId}/automations`)
    return res.data
  },

  get: async (projectId: number, id: number): Promise<AutomationRule> => {
    const res = await api.get(`/projects/${projectId}/automations/${id}`)
    return res.data
  },

  create: async (projectId: number, data: AutomationCreate): Promise<AutomationRule> => {
    const res = await api.post(`/projects/${projectId}/automations`, data)
    return res.data
  },

  update: async (projectId: number, id: number, data: Partial<AutomationCreate>): Promise<AutomationRule> => {
    const res = await api.put(`/projects/${projectId}/automations/${id}`, data)
    return res.data
  },

  delete: async (projectId: number, id: number): Promise<void> => {
    await api.delete(`/projects/${projectId}/automations/${id}`)
  },

  execute: async (projectId: number, id: number, issueId: number, context?: Record<string, any>): Promise<any> => {
    const res = await api.post(`/projects/${projectId}/automations/${id}/execute`, {
      issue_id: issueId,
      context: context || {},
    })
    return res.data
  },

  getExecutionHistory: async (issueId: number, limit?: number): Promise<AutomationExecution[]> => {
    const params = new URLSearchParams()
    if (limit) params.append('limit', limit.toString())
    const res = await api.get(`/issues/${issueId}/automation-history?${params.toString()}`)
    return res.data
  },

  getRuleExecutionHistory: async (ruleId: number, params?: { limit?: number; offset?: number; startTime?: string; endTime?: string }): Promise<AutomationExecutionResponse> => {
    const query = new URLSearchParams()
    if (params?.limit) query.append('limit', params.limit.toString())
    if (params?.offset) query.append('offset', params.offset.toString())
    if (params?.startTime) query.append('start_time', params.startTime)
    if (params?.endTime) query.append('end_time', params.endTime)
    const res = await api.get(`/automations/${ruleId}/execution-history?${query.toString()}`)
    return res.data
  },

  getProjectExecutionHistory: async (projectId: number, params?: { limit?: number; offset?: number; startTime?: string; endTime?: string }): Promise<AutomationExecutionResponse> => {
    const query = new URLSearchParams()
    if (params?.limit) query.append('limit', params.limit.toString())
    if (params?.offset) query.append('offset', params.offset.toString())
    if (params?.startTime) query.append('start_time', params.startTime)
    if (params?.endTime) query.append('end_time', params.endTime)
    const res = await api.get(`/projects/${projectId}/automation-executions?${query.toString()}`)
    return res.data
  },
}

File: backend/internal/model/automation.go
package model

import "time"

// AutomationRule represents an automation rule for a project.
type AutomationRule struct {
	BaseModel

	Name            string     `gorm:"size:100;not null" json:"name"`
	Description     string     `gorm:"type:text" json:"description"`
	ProjectID       uint64     `gorm:"not null;index" json:"project_id"`
	WorkspaceID     uint64     `gorm:"not null;index" json:"workspace_id"`
	IsEnabled       bool       `gorm:"default:true" json:"is_enabled"`
	Sequence        int        `gorm:"default:1" json:"sequence"`
	ExecutionCount  int        `gorm:"default:0" json:"execution_count"`
	TriggerType     string     `gorm:"size:50;not null" json:"trigger_type"`
	Conditions      string     `gorm:"type:text" json:"conditions"`
	Actions         string     `gorm:"type:text" json:"actions"`
	Scope           string     `gorm:"size:20;default:'all'" json:"scope"` // all, project, specific
	ScheduleConfig  string     `gorm:"type:text" json:"schedule_config"`
	LastTriggeredAt *time.Time `json:"last_triggered_at"`
	CreatedByID     *uint64    `json:"created_by_id"`
	UpdatedByID     *uint64    `json:"updated_by_id"`
}

func (AutomationRule) TableName() string {
	return "automation_rules"
}

// AutomationRuleOverride represents a per-project override of a workspace automation rule.
type AutomationRuleOverride struct {
	BaseModel

	RuleID     uint64  `gorm:"not null;index" json:"rule_id"`
	ProjectID  uint64  `gorm:"not null;index" json:"project_id"`
	IsEnabled  *bool   `json:"is_enabled"`
	CreatedByID *uint64 `json:"created_by_id"`
	UpdatedByID *uint64 `json:"updated_by_id"`
}

func (AutomationRuleOverride) TableName() string {
	return "automation_rule_overrides"
}

// AutomationExecution represents the execution of an automation rule.
type AutomationExecution struct {
	ID           uint64    `gorm:"primaryKey" json:"id"`
	RuleID       uint64    `gorm:"not null;index" json:"rule_id"`
	IssueID      uint64    `gorm:"not null;index" json:"issue_id"`
	TriggerType  string    `gorm:"size:50;not null" json:"trigger_type"`
	ContextJSON  string    `gorm:"type:jsonb" json:"context_json"`
	ActionsTaken string    `gorm:"type:jsonb" json:"actions_taken"`
	Status       string    `gorm:"size:20;not null" json:"status"`
	Error        string    `gorm:"type:text" json:"error"`
	Duration     int64     `gorm:"default:0" json:"duration"`
	ExecutedAt   time.Time `gorm:"not null" json:"executed_at"`
	CreatedAt    time.Time `gorm:"not null" json:"created_at"`
	UpdatedAt    time.Time `gorm:"not null" json:"updated_at"`
}

func (AutomationExecution) TableName() string {
	return "automation_executions"
}
